api/jobs.go

```diff
@@ -987,23 +988,24 @@ type TaskGroupSummary struct {
 // JobListStub is used to return a subset of information about
 // jobs during list operations.
 type JobListStub struct {
-	ID                string
-	ParentID          string
-	Name              string
-	Namespace         string `json:",omitempty"`
-	Datacenters       []string
-	Type              string
-	Priority          int
-	Periodic          bool
-	ParameterizedJob  bool
-	Stop              bool
-	Status            string
-	StatusDescription string
-	JobSummary        *JobSummary
-	CreateIndex       uint64
-	ModifyIndex       uint64
-	JobModifyIndex    uint64
-	SubmitTime        int64
+	ID                       string
+	ParentID                 string
+	Name                     string
+	Namespace                string `json:",omitempty"`
+	Datacenters              []string
+	Type                     string
+	Priority                 int
+	Periodic                 bool
+	ParameterizedJob         bool
+	DispatchIdempotencyToken string
```

Comment:
I added the token to the list stub to help list tokens that are active, like in the `nomad job status` output list. It's a valuable information, but I think it can be removed if it adds too much overhead.

Reply:
I'm feeling conflicted about the prominence of idempotency token in alloc lists and such.

I envisioned a very narrow case for: when a dispatch job request fails for transient reasons (e.g. http timeout, network error), we want to ensure retrying the request doesn't result in multiple dispatches of the same job.

Also, the idempotency token should probably be a uuid like token (or maybe concatenated uuid tokens), so will not be human friendly to be included in the list.

Reply:
Yeah, I wasn't sure of this one, but it wasn't a big change, so I started with it. I can remove it from the list stub 👍 

> Also, the idempotency token should probably be a uuid like token (or maybe concatenated uuid tokens), so will not be human friendly to be included in the list.

From the original issue it sounded like the token would be user-generated 🤔 

If it were a UUID I think the API implementation would have to auto-generate one at request time, no?

Reply:
Good question. It's user generated and hopefully auto-generated by the component responsible for retries. As nomad/api package doesn't do retries, it's not very appropriate for it to set one (well, it can set one automatically but it doesn't add benefit).

At the risk of mainsplaining or being condescending, i'll explain how I understand idempotency tokens to work in API system. Consider a system that attempts to dispatch a job in Nomad but got an HTTP timeout request. The dispatcher is uncertain if the connection timed out while it sending the request before the server actually processed it (hence dispatch job not starting), or it timed out after the server already dispatched the job but before the response was received. In this case, resubmitting the job can result into duplicate jobs! To avoid this scenario, the dispatcher should generate a unique idempotency token and send it with the first request and all retries. When nomad handles the retry request, Nomad avoids starting a new job if it already started one and simply return the existing job info. Does this match your understanding too?